### PR TITLE
refactor transaction manager

### PR DESCRIPTION
**Pull request by kdembler**
Should fix problems with callbacks and transactions once and for all.

Changed the approach a bit, `TransactionManager` isn't a context anymore, it just manages the display of `TransactionDialog` and monitor state of query node sync, calling callback if needed.

All the other stuff is happening inside an unstateful hook now so there's no chance of callbacks clashing/not resetting properly between separate transactions.

### CODE REVIEW

File: src/providers/transactionManager/store.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  addCallback: (callback: TransactionSyncCallback) => void
+  removeOldCallbacks: (currentBlock: number) => void
+  setDialogStep: (step: TransactionDialogStep) => void
+}
+
+export const useTransactionManagerStore = createStore<TransactionManagerStoreState, TransactionManagerStoreActions>({
+  state: { syncCallbacks: [], dialogStep: null },
+  actionsFactory: (set) => ({
+    addCallback: (callback) =>
+      set((draft) => {
```
Comment (drillprop) [inline src/providers/transactionManager/store.ts]: Why it's called `draft`? Why not just `state`?

File: src/providers/transactionManager/store.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import { ExtrinsicStatus } from '@/joystream-lib'
+import { createStore } from '@/store'
+
+type TransactionSyncCallback = {
+  targetBlock: number
+  callback: () => void
+}
```
Comment (drillprop) [inline src/providers/transactionManager/store.ts:7]: Maybe it's just a preference but don't like how it's called, I would expect only a callback from this type. It's a little bit confusing. Also, I've noticed that you doing something like this: 
```tsx
cb.callback()
``` 
which is also in my opinion non-intuitive

File: src/providers/transactionManager/useTransaction.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            } catch (e) {
+              Logger.error('Failed transaction sync callback', e)
+            }
+          }
+          resolve()
+        }
+        addCallback({ callback: syncCallback, targetBlock: block })
+      })
+
+      return new Promise<boolean>((resolve) => {
```
Comment (drillprop) [inline src/providers/transactionManager/useTransaction.ts]:
```suggestion
      return new Promise((resolve) => {
```
I don't think you need this

**Review by drillprop — APPROVED** [review]
Overall great work, much-needed PR ❤️  
Did several uploads with no problems, so it should resolve the issue.

File: src/providers/transactionManager/store.ts (hunk 1 above)
Comment (kdembler) [reply src/providers/transactionManager/store.ts]: That's the default naming that `immer` uses but changed that to `state` now.

File: src/providers/transactionManager/store.ts (hunk 2 above)
Comment (kdembler) [reply src/providers/transactionManager/store.ts:7]: Hmm yeah I just couldn't come up with better names. Changed that now, hopefully it's better

File: src/providers/transactionManager/useTransaction.ts (hunk 3 above)
Comment (kdembler) [reply src/providers/transactionManager/useTransaction.ts]: removed :+1:

File: src/providers/transactionManager/store.ts (hunk 2 above)
Comment (drillprop) [reply src/providers/transactionManager/store.ts:7]: Yep, I think it's much better now  👍🏻

**Review by drillprop — APPROVED** [review]

**Pull request merged**